### PR TITLE
gh-120 builder for generative openai module

### PR DESCRIPTION
**Pull request by shreyanshshah27**

### CODE REVIEW

**Comment by shreyanshshah27** [comment]
Hey @dirkkul 

I've tried to implement the builder for the generative-openai module. I've also added a few test cases to verify the feature. Integration test is failing right now because the `OPENAI_APIKEY` hasn't been added to the secrets yet. It would be great if you can add this key to the secrets and re-run the tests.

Also let me know your thoughts on this builder implementation. If this looks good then I'll add the remaining few integration tests for this builder.

Also in the docs it is mentioned that this would work with `Get{}` only, but as far as I understand this can work with `MultiClassBuilder` also.

**Comment by dirkkul** [comment]
Hi @shreyanshshah27 

thanks a lot for this! I didn't have time to review this, will try tomorrow.

I don't have an OpenAI key that I can put into the CI yet, will try to organize one. But could you redo the tests, so they run without that key and just skip tests that need one? I did something similar for the authentication tests [here](https://github.com/weaviate/weaviate-go-client/blob/4921d545c6e8c0bf960400a0302b22b21efbe7f8/test/auth/auth_test.go#L44).

PS if you install pre-commit gofumpt is run on every commit and your files are reformated automatically.

**Comment by shreyanshshah27** [comment]
@dirkkul No worries!! Please take your time. As you suggested, I've fixed the generative-module integration tests. But I'm facing some issue while installing gmfumpt, will try to get it resolved before next commit xD.

**Comment by dirkkul** [comment]
great thank you!

I added the OPENAI_APIKEY env variable. The tests won't run for you as secrets aren't shared with external contributors, but I'll test it locally. And after merging it will run :)

If you need any help with gofumpt, you can also ping me on slack

File: test/testsuit/generics.go [hunk 1]
```diff
@@ -150,20 +150,26 @@ func CleanUpWeaviate(t *testing.T, client *weaviate.Client) {
 // CreateTestClient running on local host 8080
 func CreateTestClient(port int, connectionClient *http.Client) *weaviate.Client {
 	integrationTestsWithAuth := os.Getenv("INTEGRATION_TESTS_AUTH")
+	openAIApiKey := os.Getenv("OPENAI_APIKEY")
 	var cfg *weaviate.Config
 	wcsPw := os.Getenv("WCS_DUMMY_CI_PW")
```
[the remark this answers is not in the document]
Comment (shreyanshshah27) [reply test/testsuit/generics.go:154]: Fixed.

File: test/testsuit/generics.go [hunk 2]
```diff
@@ -150,20 +150,26 @@ func CleanUpWeaviate(t *testing.T, client *weaviate.Client) {
...
+				Scheme:  "http",
+				Headers: map[string]string{"X-OpenAI-Api-Key": openAIApiKey},
+			}
 		}
 	} else {
 		cfg = &weaviate.Config{
 			Host:             "localhost:" + fmt.Sprint(port),
 			Scheme:           "http",
 			ConnectionClient: connectionClient,
+			Headers:          map[string]string{"X-OpenAI-Api-Key": openAIApiKey},
```
[the remark this answers is not in the document]
Comment (shreyanshshah27) [reply test/testsuit/generics.go]: Fixed.

File: weaviate/graphql/generativesearch.go [hunk 3]
```diff
@@ -0,0 +1,59 @@
+package graphql
```
[the remark this answers is not in the document]
Comment (shreyanshshah27) [reply weaviate/graphql/generativesearch.go:1]: Incorporated this change.

File: weaviate/graphql/generativesearch.go [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+func NewGSWithSingleResult(prompt string) GenerativeSearch {
+	return &singleResult{prompt: prompt}
+}
+
+func NewGSWithGroupedResult(task string) GenerativeSearch {
+	return &groupedResult{task: task}
+}
+
+func (sr *singleResult) build() string {
+	sr.prompt = "\"\"\"" + sr.prompt + "\"\"\""
```
[the remark this answers is not in the document]
Comment (shreyanshshah27) [reply weaviate/graphql/generativesearch.go]: I tried without adding """, but it wasn't working.

File: weaviate/graphql/generativesearch.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (dirkkul) [reply weaviate/graphql/generativesearch.go:1]: thanks!

**Review by dirkkul — APPROVED** [review]

**Pull request merged**